hydra-node/src/Hydra/Chain/Direct/Handlers.hs

```diff
@@ -145,11 +162,52 @@ mkChain tracer queryTimeHandle wallet ctx LocalChainState{getLatest} submitTx =
         submitTx vtx
     , -- Handle that creates a draft commit tx using the user utxo.
       -- Possible errors are handled at the api server level.
-      draftTx = \utxo -> do
+      draftTx = \utxoInputs -> do
         chainState <- atomically getLatest
         case Hydra.Chain.Direct.State.chainState chainState of
-          Initial st ->
-            sequenceA $ finalizeTx wallet ctx chainState utxo <$> commit ctx st utxo
+          Initial st -> do
+            walletUtxos <- atomically getUTxO
+            let walletTxIns = fromLedgerTxIn <$> Map.keys walletUtxos
+            let userTxIns = Set.toList $ UTxO.inputSet regularUtxo
+            let matchedWalletUtxo = filter (`elem` walletTxIns) userTxIns
+            if null matchedWalletUtxo
+              then sequenceA $ finalizeTx wallet ctx chainState (regularUtxo <> scriptUtxo) <$> buildCommitScriptTx
+              else pure $ Left FailedToDraftTxWalletUtxoDetected
+           where
+            regularUtxo = foldMap (\(u, m) -> maybe u (const mempty) m) utxoInputs
+            scriptUtxoInput =
+              foldMap
+                ( \(u, m) ->
+                    case m of
+                      Just (d, r, s) -> [(u, d, r, s)]
+                      Nothing -> []
+                )
+                utxoInputs
+
+            scriptUtxo = foldMap (\(u, _, _, _) -> u) scriptUtxoInput
+            buildCommitScriptTx = do
+              commitScriptTxBody <- commitScript ctx st regularUtxo
+              let commitScriptTx =
+                    unsafeBuildTransaction $
+                      commitScriptTxBody
+                        & addTxIns scriptUtxoInfo
+              pure commitScriptTx
+            addTxIns scriptInputs bodyTx =
+              foldl'
+                (\body scriptInput -> body & addTxIn (toScriptInput scriptInput))
+                bodyTx
+                scriptInputs
+            toScriptInput (txIn, datum, redeemer, script) =
+              ( txIn
+              , BuildTxWith $
+                  ScriptWitness ScriptWitnessForSpending $
+                    mkScriptWitness script datum redeemer
+              )
```

Comment:
This indicates that we should use `ScriptWitness` already on the type signature of `draftTx`